### PR TITLE
#164198184 User should create a profile

### PR DESCRIPTION
**Pull request by fejiroofficial**
**What does this PR do?**
This is PR for building a feature that allows users to create a profile.

**Description of Task to be completed?**
- Write unit test(s)
- Write code and migration scripts
- Write Controller functions and validators to handle feature

**How should this be manually tested**
- Clone this repo and cd into it
- run `npm install` to install dependencies.
- test using mocha with `npm test`.

**What are the relevant pivotal tracker stories**
[#164198184](https://www.pivotaltracker.com/n/projects/2313263/stories/164198184)

**Screenshots**
N/A

### CODE REVIEW

File: helpers/constants.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+export default {
```
Comment (aanchirinah) [inline helpers/constants.js]: #### Nitpick

- Is everything in here relevant to the product?

File: middlewares/validateCreateProfile.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * @returns {Object}
+ *
+ * @exports validateCreateProfile
+ */
+
+const validateCreateProfile = (req, res, next) => {
+  let {
+    firstname, lastname, username, gender, bio, phone, address, image,
+  } = req.body;
+  const urlRegex = /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9]\.[^\s]{2,})/;
```
Comment (aanchirinah) [inline middlewares/validateCreateProfile.js]: #### Nitpick

- Have a helper function for this or find a way to encapsulate it out of this file

File: models/Profile.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    },
+    image: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    }
+  }, {});
+
+  ProfileSchema.associate = (models) => {
+    ProfileSchema.belongsTo(models.User, {
+      foreignKey: 'user_id',
```
Comment (aanchirinah) [inline models/Profile.js]: #### Nitpick
- So from the documentation, it will add `user_id` to the Profile model, right?

File: models/User.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ const User = (sequelize, DataTypes) => {
   /**
    * @type {Sequelize.Model}
    */
-  const UserSchema = sequelize.define('user', {
+  const UserSchema = sequelize.define('users', {
```
Comment (aanchirinah) [inline models/User.js]: #### Critical

- Model definition should be singular not plural.

File: test/integrations/routes/profile.test.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  user_id: 1,
+  username: faker.random.words(),
+  image: faker.image.imageUrl(),
+};
+
+const token = `Bearer ${jwt.sign({ user: { id: 1 } }, 'secret', { expiresIn: '24hrs' })}`;
+
+describe('API end point for User profile', () => {
+  after(() => models.Profile.destroy({ truncate: true }));
+  it('it should create profile with correct details', (done) => {
```
Comment (aanchirinah) [inline test/integrations/routes/profile.test.js]: #### Nitpick

- Your test captions should make a proper English statement,What you have now looks like this `API end point for User profile it should create profile with correct details` .

File: test/integrations/routes/profile.test.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+          'image',
+        ]);
+        expect(res.body.profile.id).to.not.be.a('string');
+        expect(res.body.profile.user_id).to.not.be.a('string');
+        expect(res.body.message).to.be.equals('Profile created successfully');
+        done();
+      });
+  });
+
+  it('it should throw an error if firstname is not supplied', (done) => {
```
Comment (aanchirinah) [inline test/integrations/routes/profile.test.js]: #### Nitpick

- Same here s previous comment, fix all.

**Comment by aanchirinah** [comment]
@fejiroofficial , Also pull latest changes before you work on the changes

File: helpers/constants.js (hunk 1 above)
Comment (fejiroofficial) [reply helpers/constants.js]: Resolved

File: middlewares/validateCreateProfile.js (hunk 2 above)
Comment (fejiroofficial) [reply middlewares/validateCreateProfile.js]: Do you mean the regex?

File: models/Profile.js (hunk 3 above)
Comment (fejiroofficial) [reply models/Profile.js]: Yes it will

File: models/User.js (hunk 4 above)
Comment (fejiroofficial) [reply models/User.js]: done

File: middlewares/validateCreateProfile.js (hunk 2 above)
Comment (aanchirinah) [reply middlewares/validateCreateProfile.js]: @fejiroofficial , yes the regex

File: test/integrations/routes/users.test.js [hunk 7]
```diff
@@ -14,167 +13,157 @@ chai.use(chaiHttp);
...
-  beforeEach(() => {
-    // reset the dummy user data
-    dummyUser = { ...DUMMY_USER };
-    // drop tables
-    Bluebird.all([
-      models.User.destroy({ truncate: true }),
-    ]);
-  });
+  describe('POST /api/v1/users', () => {
+    beforeEach(() => {
```
Comment (aanchirinah) [inline test/integrations/routes/users.test.js]: #### Nitpick
- Why are you dropping a table in `beforeEach` and not dropping tables after the tests are all done.

Comment (fejiroofficial) [reply test/integrations/routes/users.test.js]: I am not dropping tables...the before each block is for a dummy user

Comment (aanchirinah) [reply test/integrations/routes/users.test.js]: @fejiroofficial I see you're `truncating` rather

**Pull request merged**